File: 2024/11_Plutonian_Pebbles/Day11.js
// Day 11: Plutonian Pebbles
// https://adventofcode.com/2024/day/11

const fs = require('fs');

function readInput(filename) {
  return fs.readFileSync(`${filename}.txt`, 'utf8').split(' ').map(Number);
}

const memo = new Map();

function countStones(num, blinks) {
  const key = `${num},${blinks}`;
  if (memo.has(key)) {
    return memo.get(key);
  }

  let stones = 0;
  if (blinks === 0) {
    stones = 1;
  } else if (num === 0) {
    stones = countStones(1, blinks - 1);
  } else {
    const numStr = num.toString();
    const len = numStr.length;
    if (len & 1) {
      stones = countStones(2024 * num, blinks - 1);
    } else {
      const half = len >> 1;
      const leftNum = Number(numStr.slice(0, half));
      const rightNum = Number(numStr.slice(half));
      const leftStones = countStones(leftNum, blinks - 1);
      const rightStones = countStones(rightNum, blinks - 1);
      stones = leftStones + rightStones;
    }
  }

  memo.set(key, stones);
  return stones;
}

function part1(nums) {
  return nums.reduce((totalStones, num) => totalStones + countStones(num, 25), 0);
}

function part2(nums) {
  return nums.reduce((totalStones, num) => totalStones + countStones(num, 75), 0);
}

var filename = 'Day11';
console.log('Part 1:', part1(readInput(filename))); // 55312
console.time('Part 1');
console.log('Part 1 Test:', part1(readInput(`${filename}_test`))); // 239714
console.timeEnd('Part 1');

console.log('Part 2:', part2(readInput(filename))); // 65601038650482
console.time('Part 2');
console.log('Part 2 Test:', part2(readInput(`${filename}_test`))); // 284973560658514
console.timeEnd('Part 2');

File: 2024/11_Plutonian_Pebbles/Day11.kt
// Day 11: Plutonian Pebbles
// https://adventofcode.com/2024/day/11

import java.io.File

fun readInput(filename: String) = File("$filename.txt").readText().trim().split(" ").map { it.toLong() }

val memo = mutableMapOf<Pair<Long, Int>, Long>()

fun countStones(num: Long, blinks: Int): Long {
    val key = num to blinks
    memo[key]?.let { return it }

    val stones = when {
        blinks == 0 -> 1L
        num == 0L -> countStones(1L, blinks - 1)
        else -> {
            val numStr = num.toString()
            val len = numStr.length
            if (len % 2 != 0) {
                countStones(2024L * num, blinks - 1)
            } else {
                val half = len / 2
                val leftNum = numStr.substring(0, half).toLong()
                val rightNum = numStr.substring(half).toLong()
                countStones(leftNum, blinks - 1) + countStones(rightNum, blinks - 1)
            }
        }
    }

    memo[key] = stones
    return stones
}

fun part1(nums: List<Long>): Long {    
    return nums.sumOf { countStones(it, 25) }
}

fun part2(nums: List<Long>): Long {    
    return nums.sumOf { countStones(it, 75) }
}

fun main() {
    val input = readInput("Day11")
    val inputTest = readInput("Day11_test")

    println("Part 1: ${part1(input)}") // 55312
    println("Part 1 Test: ${part1(inputTest)}") // 239714

    println("Part 2: ${part2(input)}") // 65601038650482
    println("Part 2 Test: ${part2(inputTest)}") // 284973560658514
}
